Dynamic config preview (#5252)
circleci/circleci-docs: jekyll/_cci2/dynamic-config.md

---
layout: classic-docs
title: "Dynamic Configuration"
description: "Docs page on using Setup Workflows for Dynamic Configuration"
version:
- Cloud
---

You may find that instead of manually creating each and every individual CircleCI configuration per project,

> you would prefer to generate these configurations dynamically depending on specific [pipeline]({{ site.baseurl }}/2.0/concepts/#pipelines)

Reviewer: ```suggestion
you would prefer to generate these configurations dynamically, depending on specific [pipeline]({{ site.baseurl }}/2.0/concepts/#pipelines)
```

> you would prefer to generate these configurations dynamically depending on specific [pipeline]({{ site.baseurl }}/2.0/concepts/#pipelines) 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1

parameters or file-paths.

This becomes particularly useful in cases where your team is using a monorepo, or a single repository, as opposed to

> using multiple repositories to store your code. In the case of using a monorepo, it is of course optimal to only trigger 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1

> specific builds in specific areas of your project. Otherwise, all of your microservices/project will go through 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1

> the entirety of your build, test, and deployment processes when any single update is introduced. 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1

In both of these (and many other) use cases, automatic, dynamic generation of your configuration files will optimize your
CircleCI experience and save your team both time and money.

CircleCI's dynamic configuration feature uses a `setup workflow` configuration. A `setup workflow` can contain jobs that

> `setup` children pipelines through computed pipeline parameters, or by generating followup pipelines via pre-existing scripts.

Reviewer: ```suggestion
`setup` children pipelines through computed pipeline parameters, or by generating follow-up pipelines via pre-existing scripts.
```

These computed pipeline parameters and/or generated `config.yaml` files can then be passed into an additional `config.yaml`
that potentially exist in outside directories.

In summary, CircleCI's Dynamic Configuration allows you to:

- Execute conditional workflows/commands
- Pass pipeline parameter values and/or generate additional configuration
- Trigger separate `config.yml` configurations which exist outside the default parent `.circleci/` directory

> To use our Dynamic Configuration feature, you can add the key `setup` with a value of `true` to the top-level of your 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1

> parent configuration file (in the `.circleci/` directory). This will designate that configuration as a `setup workflow` 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1

configuration.

> See the [Getting Started](#getting-started-with-dynamic-config-in-circleci) section below for more 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1

information.

## Getting Started with Dynamic Config in CircleCI
{: #getting-started-with-dynamic-config-in-circleci }

> To get started with Dynamic Config in CircleCI: 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1

- Select the project you are interested in, in the **Projects** dashboard from the CircleCI application.
- Click the **Project Settings** button in the upper-right corner.
- On the left-hand panel, select **Advanced**.
- Towards the bottom, toggle the switch for **Enable dynamic config using setup workflows** to the "on" position, as shown below:

![Enable dynamic config in the UI]({{ site.baseurl }}/assets/img/docs/dynamic-config-enable.png)

> Now, your project has the ability to dynamically generate and update configuration.

Reviewer: ```suggestion
Now, your project has the ability to dynamically generate and update a configuration.
```

> When using Dynamic Configuration, at the end of the `setup workflow`, a `continue` job from the [`continuation`](https://circleci.com/developer/orbs/orb/circleci/continuation) 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1

[`orb`]({{ site.baseurl }}/2.0/orb-intro/) must be called (**NOTE:** this does not apply if you desire to conditionally execute
workflows or steps based on updates to specified files, as described in the [Configuration Cookbook]({{ site.baseurl }}/2.0/configuration-cookbook/?section=examples-and-guides#execute-specific-workflows-or-steps-based-on-which-files-are-modified) example).

For a basic example on how to use `setup workflows` for Dynamic Configuration generation, see the [Configuration Cookbook]({{ site.baseurl }}/2.0/configuration-cookbook/?section=examples-and-guides#dynamic-configuration).
Included in the cookbook are other more in-depth examples, which will be updated as this feature matures.

For a more in-depth explanation on the behind-the-scenes pipeline creation/continuation process when dynamically generating configuration,
see our [public GitHub repository](https://github.com/CircleCI-Public/api-preview-docs/blob/master/docs/setup-workflows.md#concepts).

## Dynamic Configuration FAQs
{: #dynamic-config-faqs }

### Pipeline Parameters
{: #pipeline-parameters }

**Q:** I thought pipeline parameters could only be used with the API?

> **A:** Previously, this was true. But using our Dynamic Configuration feature, you can set pipeline parameters dynamically, 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1

- Execute conditional workflows/commands
- Pass pipeline parameter values and/or generate additional configuration
- Trigger separate `config.yml` configurations which exist outside the default parent `.circleci/` directory

To use our dynamic configuration feature, you can add the key `setup` with a value of `true` to the top-level of your 
parent configuration file (in the `.circleci/` directory). This will designate that `config.yaml` as a `setup workflow` 
configuration, enabling you and your team to get up and running with dynamic configuration.

See the [Getting started](#getting-started-with-dynamic-config-in-circleci) section below for more 
information.

## Getting started with dynamic config in CircleCI
{: #getting-started-with-dynamic-config-in-circleci }

To get started with Dynamic Config in CircleCI: 

- Select the project you are interested in, in the **Projects** dashboard from the CircleCI application.
- Click the **Project Settings** button in the upper-right corner.
- On the left-hand panel, select **Advanced**.
- Towards the bottom, toggle the switch for **Enable dynamic config using setup workflows** to the "on" position, as shown below:

>   

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 2

![Enable dynamic config in the UI]({{ site.baseurl }}/assets/img/docs/dynamic-config-enable.png)

Now, your project has the ability to dynamically generate and update configuration.

> When using dynamic configuration, at the end of the `setup workflow`, a `continue` job from the [`continuation`](https://circleci.com/developer/orbs/orb/circleci/continuation) 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1

[`orb`]({{ site.baseurl }}/2.0/orb-intro/) must be called (**NOTE:** this does not apply if you desire to conditionally execute
workflows or steps based on updates to specified files, as described in the [Configuration Cookbook]({{ site.baseurl }}/2.0/configuration-cookbook/?section=examples-and-guides#execute-specific-workflows-or-steps-based-on-which-files-are-modified) example).

For a basic example on how to use `setup workflows` for dynamic configuration generation, see the [Configuration Cookbook]({{ site.baseurl }}/2.0/configuration-cookbook/?section=examples-and-guides#dynamic-configuration).
Included in the cookbook are other more in-depth examples, which will be updated as this feature matures.

For a more in-depth explanation on the behind-the-scenes pipeline creation/continuation process when using CircleCI's dynamic configuration,
see our [public GitHub repository](https://github.com/CircleCI-Public/api-preview-docs/blob/master/docs/setup-workflows.md#concepts).

## Dynamic configuration FAQs
{: #dynamic-config-faqs }

### Pipeline parameters
{: #pipeline-parameters }

**Q:** I thought pipeline parameters could only be used with the API?

> **A:** Previously, this was true. But using our dynamic configuration feature, you can set pipeline parameters dynamically, 

Reviewer 2: MD009/no-trailing-spaces Trailing spaces Expected: 0 or 2; Actual: 1
